File: party-backend/src/main/java/cn/michaelwang/party/controller/GanDengYanController.java
package cn.michaelwang.party.controller;

import cn.michaelwang.party.domain.Game;
import cn.michaelwang.party.domain.Player;
import cn.michaelwang.party.domain.Record;
import cn.michaelwang.party.service.IGanDengYanService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/GanDengYan")
public class GanDengYanController {
    private final IGanDengYanService service;

    @Autowired
    public GanDengYanController(IGanDengYanService service) {
        this.service = service;
    }

    @GetMapping("/AddPlayer/{playerName}")
    public Map<String, List<Player>> addPlayer(@PathVariable String playerName) {
        return service.addPlayer(playerName);
    }

    @GetMapping("/StartGame")
    public Game startGame() {
        return service.startGame();
    }

    @PostMapping("/ChangePlayers")
    public Map<String, List<Player>> changePlayers(@RequestBody List<String> inGamePlayers) {
        return service.changePlayers(inGamePlayers);
    }

    @GetMapping("/CurrentGame")
    public Game currentGame() {
        return service.getCurrentGame();
    }

    @PostMapping("/AddRecord/{bang}")
    public RecordResult addRecord(@RequestBody Map<String, Integer> rawCards, @PathVariable int bang) {
        Record record = service.addRecord(rawCards, bang);
        List<Player> inGamePlayers = service.getCurrentGame().getInGamePlayers();
        List<Player> waitingPlayers = service.getCurrentGame().getWaitingPlayers();

        return new RecordResult(record, inGamePlayers, waitingPlayers);
    }
}

File: party-front/src/assets/GanDengYan/GanDengYanController.js
"use strict";

import Game from "assets/GanDengYan/Game"

module.exports = class GanDengYanController {
    constructor(http) {
        this.$http = http;
    }

    gameOver(game, success) {
        if (game.winner !== null) {
            let winnerMessage = game.winner + " wins!";
            let historyItem = {
                date: new Date().toDateString(),
                game: game
            };

            this.$http.get('party/webapi/GanDengYan/StartGame').then(response => {
                let data = response.body;
                let newGame = new Game(data.inGamePlayers, data.waitingPlayers, data.records, data.id);
                success(winnerMessage, historyItem, newGame);
            }, GanDengYanController.printError);
        }
    }

    addNewPlayer(playerName, playerSetter) {
        let retriever = new PlayerRetriever(playerSetter);
        this.$http.get('party/webapi/GanDengYan/AddPlayer/' + playerName).then(
            retriever.retrievePlayersFromResponse.bind(retriever),
            GanDengYanController.printError);
    }

    playerChanged(inGamePlayers, playerSetter) {
        let inGamePlayerNames = inGamePlayers.map(player => player.name);
        let retriever = new PlayerRetriever(playerSetter);
        this.$http.post('party/webapi/GanDengYan/ChangePlayers', inGamePlayerNames).then(
            retriever.retrievePlayersFromResponse.bind(retriever),
            GanDengYanController.printError);
    }

    static printError(response) {
        console.log("error: " + response);
    }

    record(data, recorded) {
        this.$http.post('party/webapi/GanDengYan/AddRecord/' + data.bang, data.record).then(
            response => {
                let data = response.body;
                recorded(data.record, data.inGamePlayers, data.waitingPlayers);
            }, GanDengYanController.printError
        );
    }

    getCurrentGame(gameReceived) {
        this.$http.get("/party/webapi/GanDengYan/CurrentGame").then(
            response => {
                let data = response.data;
                gameReceived(new Game(data.inGamePlayers, data.waitingPlayers, data.records, data.id));
            }, GanDengYanController.printError);
    }
};

class PlayerRetriever {
    constructor(playerSetter) {
        this._playerSetter = playerSetter;
    }

    retrievePlayersFromResponse(response) {
        let data = response.body;
        this._playerSetter(data.inGamePlayers, data.waitingPlayers);
    }
}